azure-kusto-data/azure/kusto/data/security.py

```diff
@@ -52,6 +53,28 @@ def __init__(self, kcsb):
             self._authentication_method = AuthenticationMethod.aad_device_login
             self._client_id = "db662dc1-0cfe-4e1c-a843-19a68e65be58"
 
+    def __getstate__(self):
+        """we cannot serialize the AuthenticationContext"""
+        state = dict(self.__dict__)
+        ctx = self._adal_context
+        if ctx:
+            del state['_adal_context']
+            state['adal_authority_url'] = ctx.authority.url
+            state['adal_token_cache'] = ctx.cache.serialize()
+        return state
+
+    def __setstate__(self, state):
+        """reconstruct the AuthenticationContext"""
+        authority_url = state.get('adal_authority_url')
+        if authority_url:
+            del state['adal_authority_url']
+        token_cache_string = state.get('adal_token_cache')
+        if token_cache_string:
+            del state['adal_token_cache']
+        if authority_url and token_cache_string:
+            state['_adal_context'] = AuthenticationContext(authority_url, cache=TokenCache(token_cache_string))
```

Comment:
This is the critical line -- when coming back from the pickle, we reconstruct the thread-safe adal.AuthenticationContext using the original process' token cache

In our app, we want users to be able to define functions that they can run with multiprocessing.Pool.map that accept KustoClient objects for making queries. 
The simplest case would look like this: 

```python
def f(data, client):
    query = do_a_bunch_of_work_to_create_my_query(data)
    result = client.execute(query)
    return do_a_little_more_work(result)

client = KustoClient(...)  # use aad device auth
client.execute('db', 'print 1 | take 0') # run a query to force authentication
with Pool() as p:
    results = p.starmap(f, [(data, client) for data in lots_of_datas])
```